srcs/main.c

```diff
@@ -442,21 +581,296 @@ void	receive_error_message() {
+	printf("\n");
+
+
+	uint32_t	ip_header_len = ip->ip_hl << 4;
+
+	// Because when receiving a icmp error message, the data section contains the ip header + 8 bytes of it's data section
+	// which in our case is the original icmp header we sent.
+	struct icmphdr	*icmp_header = (struct icmphdr*)(ip + 1); //TODO use ip_header_len instead
+	
+	print_icmp_header(icmp_header, ip->ip_len - ip_header_len);
+}
+
+
+// flags = MSG_DONTWAIT
+// flags = MSG_DONTWAIT | MSG_ERRQUEUE
+
+bool	get_original_headers(struct icmphdr *header, uint32_t bytes_available,
+			     struct icmphdr **original_icmp, struct ip **original_ip) {
+	if (bytes_available < ICMP_MINLEN)
+		goto no_icmp;
+
+	if (bytes_available < sizeof (struct icmp))
+		goto no_original_ip;
+
+	struct icmp	*icmp = (struct icmp*)header;
+	
+	uint32_t	bytes_left = bytes_available - sizeof(struct icmp);
+	
+	*original_ip = &icmp->icmp_ip;
+
+	uint32_t	ip_hdr_len = (*original_ip)->ip_hl << 2;
+	bytes_left		   -= ip_hdr_len;
+	
+	
+	if (bytes_left < sizeof (struct icmphdr))
+		goto no_original_icmp;
+
+	*original_icmp = (struct icmphdr*)(((uint8_t*)(*original_ip)) + ip_hdr_len);
+
+	
+	return true;
+no_icmp:
+no_original_ip:
+	*original_ip = NULL;
+no_original_icmp:
+	*original_icmp = NULL;
+	return false;
+}
+
+void	receive_response() {
+	struct msghdr		msg;
+	struct sockaddr_in	source_address;
+	int			flags = MSG_DONTWAIT;
+
+	uint8_t			msg_buffer[4096]; // Way too big buffer for receving the ECHO_REPLY...
+	uint8_t			control_buffer[20480]; // net.core.optmem_max
+
+	struct iovec		msg_iov = {
+		.iov_base = msg_buffer,
+		.iov_len  = sizeof(msg_buffer),
+	};
+	
+	msg.msg_name = &source_address;
+	msg.msg_namelen = sizeof(source_address);
+
+	msg.msg_iov = &msg_iov;
+	msg.msg_iovlen = 1;
+	
+	msg.msg_control = control_buffer;
+	msg.msg_controllen = sizeof(control_buffer);
+	
+	msg.msg_flags = 0; // Could be set after call
+
+	ft_bzero(msg_buffer, sizeof(msg_buffer));
+	ft_bzero(control_buffer, sizeof(control_buffer));
+		
+
+	ssize_t	ret = recvmsg(g_socket_fd, &msg, flags);
+
+	if (-1 >= ret) {
+		if (errno == EAGAIN || errno == EWOULDBLOCK) { // Socket hasn't received packet yet.
+			return ;
+		}
+		dprintf(2, "Silently ignoring error...\n");
+		return ;
+
+	}
+
+	size_t	packet_size_left = (size_t)ret;
+
+	uint8_t source_address_hostname_str[NI_MAXHOST];
+	uint8_t	source_address_raw_str[INET_ADDRSTRLEN];
+		
+	inet_ntop(AF_INET, &source_address.sin_addr,
+		  (char *)source_address_raw_str, sizeof(source_address_raw_str));
+	getnameinfo((const struct sockaddr *)&source_address, sizeof(source_address),
+		    (char *)source_address_hostname_str, sizeof(source_address_hostname_str),
+		    NULL, 0, 0);
+
+	struct icmphdr	*icmp_header;
+	struct ip	ip_header;
+
+	uint8_t	*cursor	= msg_buffer;
+
+	uint32_t	ip_header_len = 0;
+
+	if (packet_size_left < sizeof(struct ip)) {
+		dprintf(2, "packet too short (%ld bytes) from %s\n", ret, source_address_raw_str);
+		return ;
+	}
+	
+	if (raw_socket) {
```

Comment:
This branch is unnecessary really. This fonction, for now, is only when in raw socket mode